libmamba/include/mamba/core/subdirdata.hpp

```diff
@@ -25,45 +27,69 @@
 
 namespace mamba
 {
-    struct subdir_metadata
+    class DownloadMonitor;
+
+    class MSubdirMetadata
     {
-        struct checked_at
-        {
-            bool value;
-            std::time_t last_checked;
+    public:
 
-            bool has_expired() const
-            {
-                // difference in seconds, check every 14 days
-                return std::difftime(std::time(nullptr), last_checked) > 60 * 60 * 24 * 14;
-            }
+        struct HttpMetadata
+        {
+            std::string url;
+            std::string etag;
+            std::string last_modified;
+            std::string cache_control;
         };
 
-        static tl::expected<subdir_metadata, mamba_error> from_stream(std::istream& in);
+        using expected_subdir_metadata = tl::expected<MSubdirMetadata, mamba_error>;
+
+        static expected_subdir_metadata read(const fs::u8path& file);
+        void write(const fs::u8path& file);
+        bool check_valid_metadata(const fs::u8path& file);
+
+        const std::string& url() const;
+        const std::string& etag() const;
+        const std::string& last_modified() const;
+        const std::string& cache_control() const;
+
+        bool has_zst() const;
+
+        void store_http_metadata(HttpMetadata data);
+        void store_file_metadata(const fs::u8path& file);
+        void set_zst(bool value);
+
+    private:
+
+        static expected_subdir_metadata
+        from_state_file(const fs::u8path& state_file, const fs::u8path& repodata_file);
+        static expected_subdir_metadata from_repodata_file(const fs::u8path& json);
 
-        std::string url;
-        std::string etag;
-        std::string mod;
-        std::string cache_control;
 #ifdef _WIN32
-        std::chrono::system_clock::time_point stored_mtime;
+        using time_type = std::chrono::system_clock::time_point;
 #else
-        fs::file_time_type stored_mtime;
+        using time_type = fs::file_time_type;
 #endif
-        std::size_t stored_file_size;
-        std::optional<checked_at> has_zst;
-        std::optional<checked_at> has_bz2;
-        std::optional<checked_at> has_jlap;
 
-        void store_file_metadata(const fs::u8path& path);
-        bool check_valid_metadata(const fs::u8path& path);
+        HttpMetadata m_http;
+        time_type m_stored_mtime;
+        std::size_t m_stored_file_size;
 
-        void serialize_to_stream(std::ostream& out) const;
-        void serialize_to_stream_tiny(std::ostream& out) const;
+        struct CheckedAt
+        {
+            bool value;
+            std::time_t last_checked;
```

Comment:
Would an `std::optional<std::time_t> last_checked_with_value` be quivalent to this field and `bool value;` together?
I'm asking because that way you can avoid the potentially uninitialized values.

Reply:
No because we can have CheckedAt.value explicitely set to false, meaning we should never check for a zst index file.